python/accumulation.py

```diff
@@ -0,0 +1,712 @@
+
+            enu_fix = self._gps_utils.geo2enu(lat, lon, 0)
+            ego_localization = (
+                PointCloudsAccumulator.localization_from_sensor_to_ego_frame(
+                    RigidTransform(
+                        [enu_fix.item(0), enu_fix.item(1), alt],
+                        [
+                            0,
+                            0,
+                            0,
+                        ],  # Orientation is estimated later by the history of positions
+                    ),
+                    extrinsics,
+                )
+            )
+            ego_position = ego_localization.translation()
+            # yaw is stimated by the history of positions as NavSatFix doesn't provide headings
+            yaw = self._estimate_yaw(ego_position)
+            self._latest_ego_localization = RigidTransform(ego_position, [0, 0, yaw])
+
+    def _on_extrinsics(self, topic_name, transform_stamped: TransformStamped):
+        with self._mutex:
+            frame_id = transform_stamped.child_frame_id()
+            if (
+                not self.no_localization
+                and not self.localization_frame_id
+                and topic_name == self.localization_extrinsics_topic
+            ):
+                # Great, now we know the localization frame id
+                self.localization_frame_id = frame_id
+
+            translation = transform_stamped.transform().translation()
+            rotation = transform_stamped.transform().rotation()
+            extrinsics = RigidTransform(
+                [translation.x(), translation.y(), translation.z()],
+                [rotation.w(), rotation.x(), rotation.y(), rotation.z()],
+            )
+
+            self._extrinsics[frame_id] = extrinsics
+
+    def _on_point_cloud(self, point_cloud: PointCloud2):
+        with self._mutex:
+            ego_localization = self._get_current_ego_localization()
+            if ego_localization is None:
+                # No localization yet, skip till there is localization info
+                # TODO: If no localization for too long, report an error
+                return
+
+            frame_id = point_cloud.header().frame_id()
+            if not self.no_extrinsics and frame_id not in self._extrinsics:
+                # No extrinsics yet, skip till there is extrinsics info
+                # TODO: If no extrinsics for too long, report an error
+                return
+
+            extrinsics = None if self.no_extrinsics else self._extrinsics[frame_id]
+
+            self._accumulator.accumulate(
+                frame_id,
+                point_cloud2.read_points(point_cloud),
+                ego_localization,
+                radar_extrinsics=extrinsics,
+            )
+
+        # Out of the mutex lock on purpose, so
+        # get_points_local_frame_relative/get_points_ego_relative can be called
+        # in the callback
+        if self.on_point_cloud:
+            self.on_point_cloud(self)
```

Comment:
The thread lock might not be working as intended here. I ran the following script to test the `on_pointcloud` callback functionality. `get_points_local_frame_relative` is run twice to test if the same number of points is returned.
```
import threading
import provizio_dds

max_accumulate_frames_per_radar = 5

mutex = threading.Lock()

def do_on_pointcloud(accumulator):
    with mutex:
        print(f"\n{len(accumulator.get_points_local_frame_relative())} points were accumulated")
        print(f"{len(accumulator.get_points_local_frame_relative())} points were accumulated")

# By default, relies on rt/provizio_radar_odometry as a localization data source and rt/provizio_extrinsics as sensors extrinsics data source
points_accumulator = provizio_dds.accumulation.DDSPointCloudsAccumulator(
    max_accumulate_frames_per_radar, 
    extrinsics_topics=None, 
    on_point_cloud=lambda accumulator: do_on_pointcloud(accumulator)
)
```

The pointcloud is being sent out at 2fps, so there should be plenty of time to accumulate the pointcloud twice before the next pointcloud arrives.

![image](https://github.com/provizio/provizio_dds/assets/85248356/7aaa5fff-692f-4452-850d-5ff1091dba6b)

The output is fine for the first 3 frames, but it gives different number of points after this. This could be related to `max_frames_without_filter` because the default is 3. It could also be related to the `filter` issue on line 260.

Reply:
@kevmcapro 
The 2FPS part looks a bit suspicious here indeed, but in not in the context of multithreading - actually even without any multi-threading protection it would work fine if indeed the messages were delivered that significant time part. Even more so with mutex protection. Shouldn't also have anything to do with that filter issue, as it would then have effect of the first call to `get_points_local_frame_relative` as well (as it's done on accumulation, not on getting points). So my only explanation so far is that messages arrived sooner one after another, it can be a socket delivery thing (i.e. it processed 2 messages at once - an older and then a newer one). If this guess is correct, then the rest is fine, see:
1. Accumulation is done under a mutex protection.
2. This protection is released prior to calling `on_pointcloud`. As specified in the comment above, it's done on purpose so getting of accumulated points can be done inside the callback
3. In the callback, you `accumulator.get_points_local_frame_relative()`. Inside it, the same mutex gets locked again indeed, so there will be no data race in the thread modifying data. When this function returns, there is no lock on that mutex anymore, so in the background, DDS can already be delivering the next point cloud.
4. The second call to `accumulator.get_points_local_frame_relative()` faces a blocked mutex. It waits for its release (i.e. till the new point cloud has been processed) and then it proceeds with a new accumulated value.

If the picture above is right, you would probably see 2 more 5202 right after the very last line seen. If no, then it should be something else and a full repro code (including publishing of test point clouds and localization) would be really appreciated, so I could debug it (and add a new unit test).

Reply:
Thank you, I'll investigate it

Reply:
Seems resolved with `list(filter` change. Not rerpoduced anymore